packages/timeline-state-resolver/src/devices/telemetrics.ts

```diff
@@ -0,0 +1,168 @@
+import {
+	DeviceOptionsTelemetrics,
+	DeviceType,
+	Mappings,
+	StatusCode,
+	TelemetricsOptions,
+	TimelineObjTelemetrics,
+} from 'timeline-state-resolver-types'
+import { TimelineState } from 'superfly-timeline'
+import { Device, DeviceStatus } from './device'
+import { Socket } from 'net'
+import * as _ from 'underscore'
+import { DoOnTime } from '../doOnTime'
+import Timer = NodeJS.Timer
+
+const TELEMETRICS_NAME = 'Telemetrics'
+const SOCKET_PORT = 5000
+const TIMEOUT_IN_MS = 2000
+
+/**
+ * Connects to a Telemetrics Device on port 5000 using a TCP socket.
+ * This class uses a fire and forget approach.
+ */
+export class TelemetricsDevice extends Device<DeviceOptionsTelemetrics> {
+	private doOnTime: DoOnTime
+
+	private socket: Socket
+	private statusCode: StatusCode = StatusCode.UNKNOWN
+	private errorMessage: string
+
+	private recentlySendCalls: Map<string, number> = new Map()
+	private retryConnectionTimer: Timer | undefined
+
+	constructor(deviceId: string, deviceOptions: DeviceOptionsTelemetrics, getCurrentTime: () => Promise<number>) {
+		super(deviceId, deviceOptions, getCurrentTime)
+
+		this.doOnTime = new DoOnTime(() => this.getCurrentTime())
+		this.handleDoOnTime(this.doOnTime, 'telemetrics')
+	}
+
+	get canConnect(): boolean {
+		return true
+	}
+
+	clearFuture(_clearAfterTime: number): void {
+		// No state to handle - we use a fire and forget approach
+	}
+
+	get connected(): boolean {
+		return this.statusCode === StatusCode.GOOD
+	}
+
+	get deviceName(): string {
+		return `${TELEMETRICS_NAME} ${this.deviceId}`
+	}
+
+	get deviceType() {
+		return DeviceType.TELEMETRICS
+	}
+
+	getStatus() {
+		const messages: string[] = []
+
+		switch (this.statusCode) {
+			case StatusCode.GOOD:
+				this.errorMessage = ''
+				messages.push('Connected')
+				break
+			case StatusCode.BAD:
+				messages.push('No connection')
+				break
+			case StatusCode.UNKNOWN:
+				this.errorMessage = ''
+				messages.push('Not initialized')
+				break
+		}
+
+		if (this.errorMessage) {
+			messages.push(this.errorMessage)
+		}
+
+		const deviceStatus: DeviceStatus = {
+			statusCode: this.statusCode,
+			messages,
+			active: this.isActive,
+		}
+		return deviceStatus
+	}
+
+	handleState(newState: TimelineState, mappings: Mappings): void {
+		super.onHandleState(newState, mappings)
+		_.each(newState.layers, (timelineObject, _layerName) => {
+			const telemetricsObject: TimelineObjTelemetrics = timelineObject as unknown as TimelineObjTelemetrics
+			telemetricsObject.content.presetShotIdentifiers.forEach((presetShotIdentifier: number) => {
+				const command = `P0C${presetShotIdentifier}\r`
+
+				if (this.hasCommandBeenSendWithinLastSecond(command, newState.time)) {
+					return
+				}
+
+				this.recentlySendCalls.set(command, newState.time)
+
+				this.doOnTime.queue(newState.time, undefined, () => {
+					this.socket.write(command)
+				})
```

Comment:
I agree we shouldn't rely on blueprints making pieces that lives for less than a second.
I do think that starting to store state and look at the difference would be overkill for this device since we don't have to send commands to the device when we no longer want the shot.

I've updated the implementation to store the last received timelineObjectId (which kinda is state) and then only send a command if it receives a new timelineObjectId. This way the timelineObject could be on the timeline indefinately and we would still only send the comman once.
And since the timelineObject gets a new id every time the same Part is taken, we would send the command again, if they retake a Part (which I think would be expected behaviour)

What are your thoughts on this approach?

Reply:
Please make this integration use state, because now if we have two layer mappings, you'll again see the commands being sent every time `handleState` is called. Sample timeline:
```
literal<TimelineObjTelemetrics>({
	id: 't0',
	enable: {
		start: Date.now(),
		duration: 20000,
	},
	content: {
		deviceType: DeviceType.TELEMETRICS,
		presetShotIdentifiers: [1],
	},
	layer: 'robo0',
}),
literal<TimelineObjTelemetrics>({
	id: 't1',
	enable: {
		start: Date.now(),
		duration: 20000,
	},
	content: {
		deviceType: DeviceType.TELEMETRICS,
		presetShotIdentifiers: [2],
	},
	layer: 'robo1',
}),
```

Reply:
Also your solution seems to be prone to other edge cases. `handleState` might be called ahead of time (hence we queue the commands), but if between resolving ahead of time and the time of planned execution of a command, a new timeline arrives due to user starting an adlib on an unrelated layer, you're queue will get cleared, but probably a new command won't be queued because the Id will be identical to what you've stored in `lastReceivedCommandId`, even though that command was not sent (not sure how that is affected by partial timeline resolving, though).

Reply:
You have convinced me. 
Updated to implement DeviceWithState